test/runtests.jl

```diff
@@ -19,6 +19,23 @@ end
     @test constructorof(Tuple{Nothing, Missing})(1.0, 2) === (1.0, 2)
 end
 
+@testset "PropertyNames" begin
+    abprops = ConstructionBase.PropertyNames{(:a, :b)}()
+    @test @inferred(abprops[1]) === :a
+    p1, s = iterate(abprops)
+    @test p1 == :a
+    p2, s = iterate(abprops, s)
+    @test p2 == :b
+    @test iterate(abprops, s) === nothing
+    # using PropertyNames helps with inferrence here
+    x = AB{Union{Int,Float64},Int}(1, 2);
+    @test @inferred(ConstructionBase.PropertyNames(x)) === abprops
+    @test @inferred(setproperties(x, (a = 2, b = 3))) == AB(2, 3)
+
+    x = (a = 1, b  = 2, c = 3, d = 4)
+    @test @inferred(getproperties(x, abprops)) === (a = 1, b = 2)
+end
+
```

Comment:
This is maybe a bit redundant and pedantic, but can we add:
```suggestion
@testset "Issue #55" begin
    struct Issue55
        x::Int
        y::Union{Nothing, Int}
    end 
    function update(x)
        setproperties(x, (; x = x.x + 1))
    end
    @inferred update(Issue55(1,2))
end
```